crates/handlers/src/rate_limit.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{net::IpAddr, sync::Arc};
+
+use governor::{clock::QuantaClock, state::keyed::DashMapStateStore, Quota, RateLimiter};
+use mas_data_model::User;
+use nonzero_ext::nonzero;
+use ulid::Ulid;
+
+const PASSWORD_CHECK_FOR_REQUESTER_QUOTA: Quota = Quota::per_minute(nonzero!(3u32));
```

Comment:
Not really sure either way but 3/min seems a bit harsh, just thinking about e.g. the company's shared office which has a shared IPv4.

Reply:
I'll do config and tweak those numbers in another PR